### PR TITLE
Various cleanups + timer fixes and optimizations 

### PR DESCRIPTION
**Pull request by ghost**
This PR has improved the netperf TCP_RR performance with 50% on my system (from ~33K to ~50K). 

While analyzing the performance issues and the improvements from #160 I have noticed that there are a lot of 0 delta timer operations scheduled. On further analysis this seems to be cause by the round operations in the timer subsystem caused by the low frequency (HZ) of the clockevent driver. Setting the resolution of the clockevent driver to 1ns seems to significantly reduce the 0 delta timer operations.

@liuyuan10 could you please give this a try on your system and see if it has the same effect?

---

This change is [Reviewable](https://reviewable.io/reviews/lkl/linux/178)

### CODE REVIEW

**Comment by liuyuan10** [comment]
This is a wonderful PR. On my desktop 

CPU: Intel(R) Xeon(R) CPU E5-1650 0 @ 3.20GHz x 12
GCC: 4.8.4

branch TCP_RR(us) TCP_STREAM(Mb/s)
master: 92 420
lkl-timer-fixes: 60 655

So indeed there are 50% TCP_RR/TCP_STREAM improvement. Also CONFIG_NO_HZ_IDLE doesn't provide performance benefit any more with this PR.

File: arch/lkl/defconfig [hunk 1]
```diff
@@ -1,14 +1,11 @@
 # CONFIG_LOCALVERSION_AUTO is not set
-# CONFIG_USELIB is not set
 CONFIG_NO_HZ_IDLE=y
 # CONFIG_SYSFS_SYSCALL is not set
 CONFIG_KALLSYMS_ALL=y
 # CONFIG_BASE_FULL is not set
 # CONFIG_FUTEX is not set
-CONFIG_EPOLL=y
```
Comment (liuyuan10) [inline arch/lkl/defconfig]: I think EPOLL and EVEENTFD are needed. Will this commit remove them?

File: arch/lkl/kernel/time.c [hunk 2]
```diff
@@ -121,7 +119,7 @@ void __init time_init(void)
 	if (ret)
 		pr_err("lkl: unable to register clocksource\n");
 
-	clockevents_config_and_register(&clockevent, HZ, 0, 0xffffffff);
+	clockevents_config_and_register(&clockevent, NSEC_PER_SEC, 1, ULONG_MAX);
```
Comment (liuyuan10) [inline arch/lkl/kernel/time.c]: Among all changes in this PR, changing min_delta from 0 to 1 is actually where it 50% performance boost comes from. Is 0 min_delta a misconfiguration? I'm curious the reason behind this.

File: arch/lkl/defconfig (hunk 1 above)
Comment (opurdila) [inline arch/lkl/defconfig]: Tthey are default in newer kernels, so there is no need to explicitly mentioned them in the saved defconfig:

$ find -name Kconfig | xargs grep -A2 EPOLL
./init/Kconfig:config EPOLL
./init/Kconfig-	bool "Enable eventpoll support" if EXPERT
./init/Kconfig-	default y

File: arch/lkl/kernel/time.c (hunk 2 above)
Comment (opurdila) [inline arch/lkl/kernel/time.c]: I don't think a 0 min_delta is the issue but a combination of min and frequency. When I debugged this, the issue was that even if the delta in clockevents_program_event() was non zero, after applying the min/max/mult/shift transformation the resulting clc was zero.

Comment (liuyuan10) [inline arch/lkl/kernel/time.c]: Thanks for the explanation. Why passing a zero clc to timer_set_oneshot can cause performance drop? Does it largely increase the number of timer ticks?

**Comment by liuyuan10** [comment]
LGTM. Wonderful PR again.

File: arch/lkl/kernel/time.c (hunk 2 above)
Comment (opurdila) [inline arch/lkl/kernel/time.c]: Yes. What seems to happen is that actual delta is say 1032ns but clc is set to 0. The host timer will be set to 1ns so the timer irq will be shot very fast and the new delta will be say 980 but the clc will be again 0 which again will short a very fast timer irq, and so on.

**Pull request merged**